### PR TITLE
ProfileID, AuthV2

### PR DESCRIPTION
**Pull request by demonsh**
Profiles, AuthV2. 
Please review only AuthV2

### CODE REVIEW

**Review by vmidyllic — APPROVED** [review]

File: circuits/lib/authV2.circom [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    calcProfile.in <== userGenesisID;
+    calcProfile.nonce <== nonce;
+
+    component isSaltZero = IsZero();
+    isSaltZero.in <== nonce;
+
+    component selectProfile = Mux1();
+    selectProfile.s <== isSaltZero.out;
+    selectProfile.c[0] <== calcProfile.out;
+    selectProfile.c[1] <== userGenesisID;
```
Comment (AndriianChestnykh) [inline circuits/lib/authV2.circom:110]: You duplicated the same code in the `verifyCredentialSubjectProfile` template. Can be separated to utils

File: circuits/lib/utils/claimUtils.circom [hunk 2]
```diff
@@ -147,6 +147,36 @@ template verifyCredentialSubject() {
...
+	signal input nonce;
+
+	component header = getClaimHeader();
+	for (var i=0; i<8; i++) { header.claim[i] <== claim[i]; }
+
+	component subjectOtherIden = getClaimSubjectOtherIden();
+	for (var i=0; i<8; i++) { subjectOtherIden.claim[i] <== claim[i]; }
+
+
+	/* ProfileID calculation */
```
Comment (AndriianChestnykh) [inline circuits/lib/utils/claimUtils.circom:164]: Something wrong with the tabulation here

File: circuits/lib/utils/idUtils.circom [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    signal output out;
+
+    component hash = Poseidon(2);
+    hash.inputs[0] <== in;
+    hash.inputs[1] <== nonce;
+
+    component genesis = TakeNBits(27*8);
+    genesis.in <== hash.out;
+
+    component oldIdParts = SplitID();
```
Comment (AndriianChestnykh) [inline circuits/lib/utils/idUtils.circom:19]: Why the name is `oldIdParts`? Sounds confusing or I missed something...

File: testvectorgen/go/auth/main.go [hunk 4]
```diff
@@ -8,60 +8,146 @@ import (
...
-	privKeyHex := "28156abe7fe2fd433dc9df969286b96666489bac508612d0e16593e944c4f69f"
-	var privKey babyjub.PrivateKey
-	if _, err := hex.Decode(privKey[:], []byte(privKeyHex)); err != nil {
-		panic(err)
+	numberOfKeys := 1
+	numberOfFirstClaimsToRevoke := 1
+	signingKeyIndex := 0
+	treeLevels := 32
+
+	useNullifier := true
```
Comment (AndriianChestnykh) [inline testvectorgen/go/auth/main.go:33]: This is not named "nullifier" anymore

File: circuits/authV2.circom [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+include "lib/authV2.circom";
+
+/*
+* The identity authorization circuit.
+* User ownership of the identity verified by signed challenge.
+* Auth claim should be in the user state and not revoked.
+* User state should be genesis or added to the global state tree (available in the smart contract).
+* The state is verified out of circuits by a verifier.
+* verification inputs:
+    - userID
```
[the remark this answers is not in the document]
Comment (demonsh) [reply circuits/authV2.circom:12]: In the verification flow it is public input, isn't it?

File: testvectorgen/go/auth/main.go (hunk 4 above)
Comment (demonsh) [reply testvectorgen/go/auth/main.go:33]: testvectorgen will be rewritten

File: circuits/lib/utils/claimUtils.circom (hunk 2 above)
Comment (demonsh) [reply circuits/lib/utils/claimUtils.circom]: fixed

File: circuits/lib/utils/idUtils.circom (hunk 3 above)
Comment (demonsh) [reply circuits/lib/utils/idUtils.circom]: renamed it do genesisIDParts

File: circuits/lib/authV2.circom [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+include "./utils/idUtils.circom";
+include "../../node_modules/circomlib/circuits/mux1.circom";
+include "../../node_modules/circomlib/circuits/comparators.circom";
+include "../../node_modules/circomlib/circuits/eddsaposeidon.circom";
+
+template AuthV2(IdOwnershipLevels, onChainLevels) {
+
+    signal input userGenesisID;
+    signal input userState;
+    signal input nonce; // random number
```
[the remark this answers is not in the document]
Comment (demonsh) [reply circuits/lib/authV2.circom:13]: changed

File: circuits/lib/authV2.circom (hunk 1 above)
Comment (demonsh) [reply circuits/lib/authV2.circom]: Move selecting profile to a separate template

**Pull request merged**